packages/flutter/lib/src/widgets/scrollable.dart

```diff
@@ -755,12 +755,46 @@ class ScrollableState extends State<Scrollable> with TickerProviderStateMixin, R
     );
   }
 
-  // Returns the delta that should result from applying [event] with axis and
-  // direction taken into account.
+  // Returns the delta that should result from applying [event] with axis,
+  // direction, and any modifier specified by the ScrollBehavior taken into
+  // account for horizontal scrollers.
   double _pointerSignalEventDelta(PointerScrollEvent event) {
-    double delta = widget.axis == Axis.horizontal
-      ? event.scrollDelta.dx
-      : event.scrollDelta.dy;
+    final TargetPlatform platform = _configuration.getPlatform(context);
+    final bool horizontal = widget.axis == Axis.horizontal;
+
+    final Set<LogicalKeyboardKey> pressed = HardwareKeyboard.instance.logicalKeysPressed;
+    final bool keyModifiersAreActive = pressed.any(_configuration.pointerAxisModifiers.contains);
+    pressed.removeAll(_configuration.pointerAxisModifiers);
+    // Ensure no other keys are currently pressed so as to not conflict with
+    // other shortcuts.
+    final bool axisIsFlipped = keyModifiersAreActive & pressed.isEmpty;
+    double delta = horizontal ? event.scrollDelta.dx : event.scrollDelta.dy;
+
+    if (horizontal & axisIsFlipped) {
+      switch(platform) {
+        case TargetPlatform.macOS:
+          // Mac already handles shift to flip the input axis, so if the
+          // modifier is currently LogicalKeyboardKey.shift (the default),
+          // ignore.
+        final bool alreadyHandled = HardwareKeyboard.instance.logicalKeysPressed.any(
+            <LogicalKeyboardKey>{
+              LogicalKeyboardKey.shiftLeft,
+              LogicalKeyboardKey.shiftRight,
+            }.contains);
+          if (alreadyHandled) {
+            break;
+          }
```

Comment:
it seems weird to me that the framework needs to know this. I can't quite wrap my head around what the implications are here. Like, if you use a physical mouse with dual-axis scroll wheels on macOS and you scroll vertically while holding shift over a horizontal scroll view, does macOS scroll horizontally? does it send us deltaX or deltaY? What about if you scroll horizontally with that device? What about if you scroll horizontally with that device while holding shift?

When does the macOS behaviour actually trigger? When I scroll up and down on my touchpad on macOS while holding shift, Chrome and Finder still scroll vertically. I don't see any settings to change this.

Reply:
I have a few more mice out on order that have balls for multidirectional input,  but here is how I have observed the behavior thus far with the collection of devices I have...

On mac, when using a discrete scrolling input device (like a mouse wheel), if the data we receive from Mac is for example `deltaX = 0, deltaY = 20`, when shift is pressed, the data we receive is `deltaX = 20, deltaY = 0`. I have not found a way to disable, alter or make Mac not do this.

It is unfortunate that Mac does this while other platforms don't. But it is something we either need to address either here or in the embedder, which based on the comment in https://github.com/flutter/flutter/issues/75180#issuecomment-1320571257:

> The embedder itself should as much as possible just pass through the OS data as completely as possible while coercing it into our protocol (so e.g. making sure down events are matched by up events and so on). 

sounded like we should handle it here. I am fine with either approach, but we need to pick one. I am really trying to push to resolve this issue rather than having it continue to sit in limbo.

> When I scroll up and down on my touchpad on macOS while holding shift, Chrome and Finder still scroll vertically. 

The touchpad is handled differently, since it is continuous input. When using a touchpad, there is a start, update, and end, just like with normal drag gestures. Trackpads take a different code path now, with a few exceptions that we're currently working out in other PRs.

Reply:
By "The embedder itself should as much as possible just pass through the OS data as completely as possible while coercing it into our protocol (so e.g. making sure down events are matched by up events and so on)" I meant that the embedder should do whatever it needs to do to make the event stream seen by the framework be as platform-neutral as possible. Meaning if macOS is doing some weird (but predictable) munging of the data, the Flutter macOS embedder should undo it, but if macOS is giving us _more_ data (e.g. saying "wheel turned vertically, scroll horizontally") then that's what we should send to the framework, and the framework could then say "well on platforms that send us an explicit scroll direction, we use that, and on others we use the keyboard data and wheel data to figure out how to scroll".

Reply:
Thanks for clarifying.

> well on platforms that send us an explicit scroll direction, we use that

I found a flag that could tell us "Mac probably flipped this" (more below), but we would still need to add special handling here if we were to honor an explicit scroll direction. Otherwise, if the Flutter developer has the same 'shift' modifier enabled, we would end up flipping it back.

Or maybe we would do nothing and just use the info provided if the OS says "this is explicit", but then it could further make the expected behavior to the Flutter developer cloudy if there are gotchas like that.

Maybe this is like us blinking our own cursor - if I can undo the manipulation Mac does in the engine, the framework can handle providing the native experience of flipping the axis via shift by default.